### ISSUE TITLE
Cancellable long-running Effect loses first value

### ISSUE THREAD

**Issue by jsmerola** [pointfreeco/swift-composable-architecture#466]
**Describe the bug**
Marking a long-running `Effect` that immediately emits a value after being subscribed to causes that value to be lost.

**To Reproduce**
[CancellableIssue.zip](https://github.com/pointfreeco/swift-composable-architecture/files/6244366/CancellableIssue.zip)

The attached project has a `Repository` effect that wraps an `NSFetchedResultsController`. The view displays the count of items in Core Data, and has a button to add another to the database. When you first launch the app, you'll see "0 names". In the `onAppear` on the view, we trigger the start of the repo. If you click the button a few times, you'll see the number increment. Do this a few times and then relaunch the app. When the view appears, you'll see it immediately says "x names", where x is the number of times you clicked the button in the last run (i.e. the number of items in Core Data).

Now go to the reducer in Core.swift and uncomment the `cancellable` line:

```swift
case .startReadingFromCoreData:
    return Repository<Thing, Int>(context: viewContext, convert: \.count)
        .map(Action.databaseUpdated)
        .eraseToEffect()
        // Uncomment this line to introduce the issue
        // .cancellable(id: "ID", cancelInFlight: true)
```

Re-run the app with this change, and when the view loads, you'll see "0 names" again. Only after clicking the button again will this update to "x+1 names". The initial value emitted by the repo was lost after making it cancellable.

**Expected behavior**
Making a long-running `Effect` cancellable should not cause the initial immediate value emitted to be lost.

**Screenshots**
https://user-images.githubusercontent.com/2407424/113312889-b90a8e00-92d8-11eb-8af3-59f7057fae33.mp4

**Environment**
 - Xcode 12.3 (12C33)
 - Swift 5
 - iOS 14.3

**Comment by erikolsson** [comment]
this is likely because you're not receiving the effect on the main queue. try changing your code to:
```
    case .startReadingFromCoreData:
        return Repository<Thing, Int>(context: viewContext, convert: \.count)
            .map(Action.databaseUpdated)
            .receive(on: DispatchQueue.main)
            .eraseToEffect()
            .cancellable(id: "ID", cancelInFlight: true)
```

**Comment by jsmerola** [comment]
You're right, that fixed it. So does `cancellable` move to a different queue? The Repository was initialized with the viewContext, which is tied to the main queue, right? So shouldn't this have already been on the main queue by default?

**Comment by RabugenTom** [comment]
Everything is on the main thread, as you can check with few assertions. The issue is elsewhere. 

I think that's because cancellable effects are behaving like a deferred `PassthroughSubject`, and they will only emit values received after subscription. The `receive(on:` is only fixing the order of calls in a way I'm not sure yet to understand, and the `Repository` is sending its initial result once the cancellable subscription is established.

This is a gotcha and there's maybe room for improvement with the API.

**Comment by fabstu** [comment]
Or is this bout [Deffered](https://heckj.github.io/swiftui-notes/#reference-deferred), used [in cancellable](https://github.com/pointfreeco/swift-composable-architecture/blob/e1bd51b1761c9f7ab4001928c73f357c9116f807/Sources/ComposableArchitecture/Effects/Cancellation.swift#L30)?

**Comment by jsmerola** [comment]
So the repository effect is being wrapped inside a Deferred when marked cancellable. But I still don't really understand why the initial value is being sent before the Deferred is being subscribed to now, since I only trigger the initial fetch after a subscription is received. Is it because the wrapped effect (the repo in this case) is being subscribed to inside of the Deferred? But shouldn't this not actually run until the Deferred itself is subscribed to?

**Comment by fabstu** [comment]
My note was only a guess, I'm unsure too. Are you return a PassthroughSubject in Repository or something like it? Here a couple passing thoughts on it. If you use a subject its probably not about the .cancellable but the issue noted by Rabu.

Not using the `Deferred` makes the `PassthroughSubject` maybe send the value before the store is subscribed, making the value lost which is usual with `PassthroughSubject`.

The `Deferred` in Repository then is only executed once the subscribe from `.cancellable` is executed, which is only executed once the store subscribes I imagine.

The `.receive(on:)` maybe works because the value sent from the `PassthroughSubject` has to wait in an DispatchQueue.main.async call to deliver the value further, which is only executed after the current execution stops, probably.

**Comment by RabugenTom** [comment]
FWIW and if I'm not mistaken, the `receive(on:)` [is sync since iOS 13.3](https://forums.swift.org/t/combine-receive-on-runloop-main-loses-sent-value-how-can-i-make-it-work/28631/39), because of similar issues. If you search Swift forums for lost values and `PassthroughSubject`, you will find a lot of references. 

`CurrentValueSubject` doesn't suffer this kind of issues, but it needs a`Output` value to be instantiated. If we replace the `PassthroughSubject` in TCA `cancellable` implementation by
```swift
final class OptionalValueSubject<Output, Failure>: Subject where Failure: Error {
    private var currentValueSubject = CurrentValueSubject<Output?, Failure>(nil)
    
    func send(_ value: Output) {
        currentValueSubject.send(value)
    }
    
    func send(completion: Subscribers.Completion<Failure>) {
        currentValueSubject.send(completion: completion)
    }
    
    func send(subscription: Subscription) {
        currentValueSubject.send(subscription: subscription)
    }
    
    func receive<S>(subscriber: S) where S : Subscriber, Failure == S.Failure, Output == S.Input {
        currentValueSubject.compactMap { $0 }.receive(subscriber: subscriber)
    }
}
```
which wraps and maps an `CurrentValueSubject<Output?, Failure>` and doesn't need a value to be instantiated ([found there](https://forums.swift.org/t/combine-receive-on-runloop-main-loses-sent-value-how-can-i-make-it-work/28631/16)), it works as expected, without needing `receive(on:)`.

I'm not saying that this is the right thing to do, as `CurrentValueSubject` are heavier than `PassthroughSubject`'s, but it shows that the issue revolves around the `PassthroughSubject` and it's instantiation. 

It's not yet clear to me why the `receive(on:)` is enough to fix the issue.

**Comment by RabugenTom** [comment]
*[I'm editing my previous comment as the `OptionalCurrentValueSubject` only fixes the current issue using `Repository`, but not the more general issue below]*

If one replaces  the `Repository` subscription by the trivial
```swift
Effect(value: Action.databaseUpdated(50))
  .cancellable(id: "ID", cancelInFlight: true)
```
Nothing is received without `receive(on:)`. Even if it makes no practical sense in this case, I'm still expecting to receive this effect. Inserting `.print().eraseToEffect()` between the `Effect` and `cancellable` shows that the subscription is active and the effect is emitted. I guess we need to go deeper to see why the reducer doesn't seem to receive this effect back.

I think that the difference of behavior with or without the unrelated `receive(on:)` would qualify the issue as a bug.
